clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/OperationResolver.java

```diff
@@ -0,0 +1,95 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client.internal.store.operations;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.ehcache.clustered.client.internal.store.operations.codecs.OperationsCodec;
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+
+public class OperationResolver<K, V> {
+
+  private final OperationsCodec<K, V> codec;
+
+  public OperationResolver(final OperationsCodec<K, V> codec) {
+    this.codec = codec;
+  }
+
+  /**
+   * Extract the {@code Element}s from the provided {@code Chain} that are not associated with the provided key
+   * and create a new {@code Chain}
+   *
+   * Separate the {@code Element}s from the provided {@code Chain} that are associated and not associated with
+   * the provided key. Create a new chain with the unassociated {@code Element}s. Resolve the associated elements
+   * and append the resolved {@code Element} to the newly created chain.
+   *
+   * @param chain a heterogeneous {@code Chain}
+   * @param key a key
+   * @return the resolved {@code Chain}
+   * @throws ClassNotFoundException
+   */
+  @SuppressFBWarnings({ "DLS_DEAD_LOCAL_STORE", "UC_USELESS_OBJECT" })
+  public Chain resolve(Chain chain, K key) {
+    List<Operation<K>> operations = new ArrayList<Operation<K>>();
+    List<Element> elements = new ArrayList<Element>();
+    for (Element element : chain) {
+      ByteBuffer payload = element.getPayload();
+      Operation<K> operation = codec.decode(payload);
+      if(key.equals(operation.getKey())) {
+        operations.add(operation);
+      } else {
+        elements.add(element);
+      }
+    }
+    Operation<K> resolvedOperation = resolve(operations);
+    ByteBuffer payload = codec.encode(resolvedOperation);
+    Element resolvedElement = null; // TODO: 13/05/16 build an element using the payload
+    elements.add(resolvedElement);
+    Chain newChain = null;  // TODO: 13/05/16  build the chain using the list of elements and the resolvedElement
+    return newChain;
+  }
+
+  /*
+    Iterate the list of operations in the reverse order and find the first PUT or REMOVE from the end.
+    The entries prior to that can be ignored as a PUT or REMOVE overrides the effect of the execution
+    of all those preceding operations.
+   */
+  Operation<K> resolve(List<Operation<K>> opList) {
+    if(opList.isEmpty()) {
+      return null;
+    }
+
+    ListIterator<Operation<K>> opsIterator = opList.listIterator(opList.size()-1);
+    Operation<K> current = opsIterator.next();
+    while (current.getOpCode() != OperationCode.PUT &&
```

Comment:
That's a nice optimization but it should be better abstracted, ie: the `Operation` itself should be able to tell if it is going to ignore the previous value or not instead of having to maintain a set of `if (opcode != ZZZ)` conditions here.